### PR TITLE
GH-14775: [Go] Fix UnionBuilder.Len implementations

### PR DESCRIPTION
**Pull request by nwt**
(*DenseUnionBuilder).Len and (*SparseUnionBuilder).Len always return 0. They should instead return the number of values appended.

This bug exists because these methods are promoted (via unionBuilder) from (*builder).Len, which reads builder.length, but DenseUnionBuilder and SparseUnionBuilder don't update that field.

Fix by adding a (*unionBuilder).Len implementation that calls unionBuilder.typesBuilder.Len.

### CODE REVIEW

**Review by zeroshade — APPROVED** [review]
Thanks for this! Once all the tests in the CI pass, I'll merge this.

**Comment by nwt** [comment]
@zeroshade: You're welcome! And thanks to you for all the work you've done in this repository!

**Comment by zeroshade** [comment]
Actually, I need to wait until the merge script is updated from #14750 before I can merge this.

**Comment by jorisvandenbossche** [comment]
@zeroshade I took the liberty to merge this as a final test of the merge script changes from https://github.com/apache/arrow/pull/14750

**Pull request merged**
